designs/0035-pre-model-gqs.md

```diff
@@ -0,0 +1,556 @@
+    if (missing1[n]) {
+      x1[n] = normal_rng(alpha1 * x2[n], sigma1);
+    }
+    if (missing2[n]) {
+      x2[n] = normal_rng(alpha2 * x1[n], sigma2);
+    }
+  }
+}
+model {
+  x1 ~ normal(alpha1 * x2, sigma1); 
+  x2 ~ normal(alpha2 * x1, sigma2);
+  y ~ normal(beta1 * x1 + beta2 * x2, sigma);
+}
+```
+
+The result will be a multiple imputation, not a coherent joint
+Bayesian model.  This will work even if the missing data is discrete
+or the conditional distributions are not coherent. 
+
+## Stochastic gradient descent
+
+The usual model in Stan evaluates the likelihood for the given data
+set and a prior.  There has, until now, never been a good way to do
+stochastic gradient descent over subsampled data.  Now we can do
+that.  This is the simplest possible example.  Suppose we start with
+this model.
+
+```stan
+data {
+  int<lower=0> N;
+  array[N] int<lower=0, upper=1> y;
+}
+parameters {
+  real<lower=0, upper=1> theta;
+}
+model {
+  theta ~ beta(5, 5);
+  y ~ bernoulli(theta);
+}
+```
+
+We can move to stochastic gradient descent by randomly subsampling `y`
+in the `latent data` block.  As we do this, it's traditional to
+scale the likelihood to match the original data size.
+
+
+```stan
+data {
+  int<lower=0> N;
+  int<lower=0> N_sub;
+  array[N] int<lower=0, upper=1> y;
+}
+transformed data {
+  real one_over_N_sub = 1.0 / N_sub;
+  real N_over_N_sub = N * one_over_N_sub;
+  simplex[N_sub] unif = rep_vector(one_over_N_sub, N_sub);
+}
+parameters {
+  real<lower=0, upper=1> theta;
+}
+latent data {
+  array[N_sub] y_sub;
+  for (n in 1:N_sub) {
+    y_sub[n] = categorical_rng(unif);
+  }
+}
+model {
+  theta ~ beta(5, 5);
+  # scale data weight to match population size
+  target += N_over_N_sub * bernoulli_lpmf(y_sub | theta);
+}
+```
+
+If you plug this model into optimization, by default it will apply
+quasi-Newton steps using stochastic gradient.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation 
+
+## Stan Reference Manual
+
+Stan's block structure and execution model is described in the
+reference manual in the
+[Program Blocks chapter](https://mc-stan.org/docs/reference-manual/blocks.html)
+and the
+[Program Execution chapter](https://mc-stan.org/docs/reference-manual/execution.html).
+This feature will add a new section to the program blocks chapter,
+which will read as follows.
+
+### Program Blocks for latent data
+
+The latent data block appears after the transformed parameters block
+but before the odel block.  Like the parameters, transformed
+parameters, and generated quantities, the latent data will be included
+in the output and the top-level variables will be available in all
+later blocks.  Latent data will also have access to the data,
+transformed data, parameters, and transformed parameters.
+
+#### Iteration number 
+
+The iteration number can be optionally set in the latent data block 
+from the outside.  There will be an `iteration_number__` parameter 
```

Comment:
@andrewgelman's been bugging us to do this for ages so that 

a.  he can write something like `if (mod(iteration_number, 100)) print(...)`, which I think is probably not going to do what he wants because it'll print every leapfrog iteration, and

b.  so we can implement something like tempering by using iteration number to modify density.  Again, all sorts of issues with these algorithms.

Reply:
This feels adjacent to the main point of the proposal. As long as we don't introduce state to the model class I'm fine with this

Reply:
I should also add a big note that this proposal does *not* include any mutable state in the model.

Yes, iteration number is adjacent and we could put it in a separate proposal, but it's going to be really easy to do this way, I think.  I can take it out if people think it's over-reaching.